osu.Server.Spectator.Tests/Multiplayer/MultiplayerCountdownTest.cs

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Moq;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Multiplayer.Countdown;
+using osu.Server.Spectator.Hubs;
+using Xunit;
+
+namespace osu.Server.Spectator.Tests.Multiplayer
+{
+    public class MultiplayerCountdownTest : MultiplayerTest
+    {
+        [Fact]
+        public async Task CannotStartCountdownIfNotReady()
+        {
+            await Hub.JoinRoom(ROOM_ID);
+
+            await Assert.ThrowsAsync<InvalidStateException>(async () => await Hub.SendMatchRequest(new StartMatchCountdownRequest { Delay = TimeSpan.FromMinutes(1) }));
+
+            using (var usage = Hub.GetRoom(ROOM_ID))
+            {
+                var room = usage.Item;
+                Debug.Assert(room != null);
+
+                Assert.Null(room.Countdown);
+                Receiver.Verify(r => r.MatchEvent(It.IsAny<CountdownChangedEvent>()), Times.Never);
+            }
+        }
+
+        [Fact]
+        public async Task GameplayStartsWhenCountdownEnds()
+        {
+            await Hub.JoinRoom(ROOM_ID);
+            await Hub.ChangeState(MultiplayerUserState.Ready);
+
+            await Hub.SendMatchRequest(new StartMatchCountdownRequest { Delay = TimeSpan.FromMinutes(1) });
+            waitForCountingDown();
+
+            using (var usage = Hub.GetRoom(ROOM_ID))
+            {
+                var room = usage.Item;
+                Debug.Assert(room != null);
+
+                Assert.NotNull(room.Countdown);
+                Assert.InRange((room.Countdown!.EndTime - DateTimeOffset.Now).Seconds, 30, 60);
+                GameplayReceiver.Verify(r => r.LoadRequested(), Times.Never);
+            }
+
+            finishCountdown();
+
+            using (var usage = Hub.GetRoom(ROOM_ID))
+            {
+                var room = usage.Item;
+                Debug.Assert(room != null);
+
+                Assert.Null(room.Countdown);
+                GameplayReceiver.Verify(r => r.LoadRequested(), Times.Once);
+            }
+        }
+
+        [Fact]
+        public async Task GameplayDoesNotStartWhenCountdownCancelled()
+        {
+            await Hub.JoinRoom(ROOM_ID);
+            await Hub.ChangeState(MultiplayerUserState.Ready);
+
+            await Hub.SendMatchRequest(new StartMatchCountdownRequest { Delay = TimeSpan.FromMinutes(1) });
+            await Hub.SendMatchRequest(new StopCountdownRequest());
+
+            finishCountdown();
```

Comment:
I don't think this should be here? Isn't the aim of this test to test the `StopCountdownRequest` part?